File: 01. Stack/[20]有效的括号.java
//给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。 
//
// 有效字符串需满足： 
//
// 
// 左括号必须用相同类型的右括号闭合。 
// 左括号必须以正确的顺序闭合。 
// 
//
// 
//
// 示例 1： 
//
// 
//输入：s = "()"
//输出：true
// 
//
// 示例 2： 
//
// 
//输入：s = "()[]{}"
//输出：true
// 
//
// 示例 3： 
//
// 
//输入：s = "(]"
//输出：false
// 
//
// 示例 4： 
//
// 
//输入：s = "([)]"
//输出：false
// 
//
// 示例 5： 
//
// 
//输入：s = "{[]}"
//输出：true 
//
// 
//
// 提示： 
//
// 
// 1 <= s.length <= 104 
// s 仅由括号 '()[]{}' 组成 
// 
// Related Topics 栈 字符串 
// 👍 2432 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    /*
    方法一：直观：
    public boolean isValid(String s) {
        if (s == null || s.length() == 0) {
            return true;
        }

        final int N = s==null ? 0:s.length();

        if ((N & 0x01) == 1) {
            return false;
        }

        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < N; i++) {
            Character cur = s.charAt(i);
            if (cur == '(' || cur == '[' || cur == '{') {
                stack.push(cur);
            } else if (cur == ')') {
                if (stack.empty() || stack.peek() != '(') {
                    return false;
                }
                stack.pop();
            } else if (cur == '}') {
                if (stack.empty() || stack.peek() != '{') {
                    return false;
                }
                stack.pop();
            } else if (cur == ']') {
                if (stack.empty() || stack.peek() != '[') {
                    return false;
                }
                stack.pop();
            }
        }
        return stack.isEmpty();
    }

     */

    // 方法二：
    public boolean isValid(String s) {
        if (s == null || s.length() == 0) {
            return true;
        }

        final int N = s.length();

        if ((N & 0x01) == 1) {
            return false;
        }

        Stack<Character> stack = new Stack<>();

        for (Character cur : s.toCharArray()) {
            if (cur == '(') {
                stack.push(')');
            } else if (cur == '[') {
                stack.push(']');
            } else if (cur == '{') {
                stack.push('}');
            } else if (stack.isEmpty() || stack.pop() != cur) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 01. Stack/[20]有效的括号.py
# 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。 
# 
#  有效字符串需满足： 
# 
#  
#  左括号必须用相同类型的右括号闭合。 
#  左括号必须以正确的顺序闭合。 
#  每个右括号都有一个对应的相同类型的左括号。 
#  
# 
#  
# 
#  示例 1： 
# 
#  
# 输入：s = "()"
# 输出：true
#  
# 
#  示例 2： 
# 
#  
# 输入：s = "()[]{}"
# 输出：true
#  
# 
#  示例 3： 
# 
#  
# 输入：s = "(]"
# 输出：false
#  
# 
#  
# 
#  提示： 
# 
#  
#  1 <= s.length <= 10⁴ 
#  s 仅由括号 '()[]{}' 组成 
#  
# 
#  Related Topics 栈 字符串 👍 3473 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        # 先判断特殊情况
        if s is None or len(s) == 0:
            return False
        if len(s) % 2 == 1:
            return False

        # 模拟栈的运行
        stack = list()
        for cur in s:
            if cur == "(" or cur == "[" or cur == "{":
                stack.append(cur)
            elif cur == ")":
                if not stack or stack.pop() != "(":
                    return False
            elif cur == "]":
                if not stack or stack.pop() != "[":
                    return False
            elif cur == "}":
                if not stack or stack.pop() != "{":
                    return False
            else:
                return False

        return not stack
# leetcode submit region end(Prohibit modification and deletion)
